### PR TITLE
[Bug fix] Navigation scenario: ensure entity placement within constrained environment boundaries

### PR DESCRIPTION
**Pull request by Giovannibriglia**
It resolves [Issue #138](https://github.com/proroklab/VectorizedMultiAgentSimulator/issues/138#issue-2497134874)

**Changes Made**:
- Removed `self.world_semidim`.
- Introduced `self.world_semidim_x` and `self.world_semidim_y`, which default to 1 (as before). If the user opts for a limited-dimension environment, these variables are set to the values of `self.x_semidim` and `self.y_semidim`.

**Enhancement**:
- Ensured that all entities are correctly placed within the defined boundaries."

### CODE REVIEW

File: vmas/simulator/environment/environment.py
Comment (matteobettini) [inline vmas/simulator/environment/environment.py:1]: Remove this

**Review by matteobettini — APPROVED** [review]
Cool! We should have caught this as part of #133, but I think it is ok since we didn't make a release in the mean time

File: vmas/simulator/environment/environment.py
Comment (Giovannibriglia) [reply vmas/simulator/environment/environment.py:1]: L'ho fatto correttamente?

**Comment by Giovannibriglia** [comment]
> Cool! We should have caught this as part of #133, but I think it is ok since we didn't make a release in the mean time

I noticed that in the discovery scenario, the implementation of the x and y bounds differs slightly from what I've done. Here's the relevant link: https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/ff5836304bf6019dddf36be78296784d25a2f9aa/vmas/scenarios/discovery.py#L136.

How would you suggest proceeding? In the navigation scenario, this pull request introduces two new variables, while in the discovery scenario, the world object is used instead.

**Comment by Giovannibriglia** [comment]
I noticed that in the discovery scenario, the implementation of the x and y bounds differs slightly from what I've done. Here's the relevant link: https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/ff5836304bf6019dddf36be78296784d25a2f9aa/vmas/scenarios/discovery.py#L136

How would you suggest proceeding? In the navigation scenario, this pull request introduces two new variables, while in the discovery scenario, the world object is used instead.

**Comment by matteobettini** [comment]
> I noticed that in the discovery scenario, the implementation of the x and y bounds differs slightly from what I've done. Here's the relevant link:
> 
> https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/ff5836304bf6019dddf36be78296784d25a2f9aa/vmas/scenarios/discovery.py#L136
> 
> How would you suggest proceeding? In the navigation scenario, this pull request introduces two new variables, while in the discovery scenario, the world object is used instead.

In discovery `self.world.x_semidim` is guaranteed to not be None, while here not. I think the current PR makes sense

**Comment by Giovannibriglia** [comment]
> > I noticed that in the discovery scenario, the implementation of the x and y bounds differs slightly from what I've done. Here's the relevant link:
> > https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/ff5836304bf6019dddf36be78296784d25a2f9aa/vmas/scenarios/discovery.py#L136
> > 
> > How would you suggest proceeding? In the navigation scenario, this pull request introduces two new variables, while in the discovery scenario, the world object is used instead.
> 
> In discovery `self.world.x_semidim` is guaranteed to not be None, while here not. I think the current PR makes sense

Great! Could you explain why the discovery scenario is boundary-limited, whereas navigation is not? What’s the reasoning behind this?

**Comment by matteobettini** [comment]
I am wondering about another thing.

If users want to specify the semidims for spawning entities but not have hard bounds, they currently can't.

What about:
- having 3 kwargs: the semidims ones that we already introduced, and a now one `enfornce_bounds` (defaults False)

So that:
- if none is specified it is like before -1,1 without bounds
- if the semidim are defined only, they rebulate the range of spawning entities but not the world semidims
- if the semidim are defined and `enfornce_bounds` is true they regulate both the world bounds and the spawning bounds

**Comment by matteobettini** [comment]
> discovery scenario is boundary-limited, whereas navigation is not? What’s the reasoning behind this?

Just the way they were implemented, which we have to keep as default for bc-compatibility. Now we have the option of bounding navigation too so they are more similar, just with different defaults

**Comment by Giovannibriglia** [comment]
> I am wondering about another thing.
> 
> If users want to specify the semidims for spawning entities but not have hard bounds, they currently can't.
> 
> What about:
> 
> * having 3 kwargs: the semidims ones that we already introduced, and a new one `enfornce_bounds` (defaults False)
> 
> So that:
> 
> * if none is specified it is like before -1,1 without bounds
> * if the semidim are defined only, they regulate the range of spawning entities but not the world semidims
> * if the semidim are defined and `enfornce_bounds` is true they regulate both the world bounds and the spawning bounds
> * if `enfornce_bounds` is true and they do not give any semidims -> error

Which scenario are you referring to?  I didn't find `enforce_bounds` variable :/

**Comment by matteobettini** [comment]
> Which scenario are you referring to? I didn't find `enforce_bounds` variable :/

I am proposing to introduce it here

**Comment by Giovannibriglia** [comment]
> > Which scenario are you referring to? I didn't find `enforce_bounds` variable :/
> 
> I am proposing to introduce it here

Oh okay, now I got you. I believe this could be a valuable enhancement. If I understand correctly:

- semidim variables are used in the entity spawning process.
- `enforce_bounds` is used to set the environment's limits when the semidim variables are not None

Therefore: 
- If both semidim and `enforce_bounds` are active: entities are spawned within the constrained world limits and the world is constrained too.
-  If semidim is active but `enforce_bounds` is not: entities are spawned within semidim's limits, but the world is unlimited.
-  If `enforce_bounds` is active but semidim is not: **error**. 
- Otherwise: the default range (-1, 1) is used for spawning, and the world is unlimited.

**Comment by matteobettini** [comment]
> Oh okay, now I got you. I believe this could be a valuable enhancement. If I understand correctly:
> 
> * semidim variables are used in the entity spawning process.
> * `enforce_bounds` is used to set the environment's limits when the semidim variables are not None
> 
> Therefore:
> 
> * If both semidim and `enforce_bounds` are active: entities are spawned within the constrained world limits and the world is constrained too.
> * If semidim is active but `enforce_bounds` is not: entities are spawned within semidim's limits, but the world is unlimited.
> * If `enforce_bounds` is active but semidim is not: **error** occurs.
> * Otherwise: the default range (-1, 1) is used for spawning, and the world is unlimited.

Exactly!

Let’s also add this comments to the kwargs explaining what they do.

Also there is the edge case of only one of the semidims been specified. In the case this happens:
- if the world is bounded, we bound only the specified one
- For spawning we default to 1 for the other

**Comment by Giovannibriglia** [comment]
Hi :)

I've been considering the solution, but I've realized that it might lead to navigation behavior that's different from other scenarios. Let me explain:

- In our proposed solution, `x_semidim` and  `y_semidim` were used as variables in the entity spawning process. However, in all other scenarios, they act as environment boundaries. Therefore, we should use these variables as follows:
```python
self.world_spawning_x = kwargs.pop("x_semidim", None)
self.world_spawning_y = kwargs.pop("y_semidim", None)
```
- Following that, there's the initialization of `enforce_bounds`:
```python
self.enforce_bounds = kwargs.pop("enforce_bounds", False)
```
I set this to False by default to maintain consistency with previous cases, but I'm open to any suggestions you might have regarding this.
- After developing this logic, here's the implementation I've come up with, which I believe aligns with our previous discussions:
```python
if self.enforce_bounds:
    if self.world_spawning_x is None or self.world_spawning_y is None:
        raise ValueError("enforce_bounds is active, but one or both semidims are None.")
    self.x_semidim = self.world_spawning_x
    self.y_semidim = self.world_spawning_y
else:
    self.x_semidim = None
    self.y_semidim = None
```
- and I will use  `world_spawning_x ` and  `world_spawning_y ` for the entities process spawning in https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/ff5836304bf6019dddf36be78296784d25a2f9aa/vmas/scenarios/navigation.py#L133 and https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/ff5836304bf6019dddf36be78296784d25a2f9aa/vmas/scenarios/navigation.py#L150

I think this approach is fairly clear because `x_semidim` and  `y_semidim`  are used to initialize `world_spawning_x` and `world_spawning_y`. What are your thoughts?

The key idea we need to ensure is: "If the environment is constrained by dimensions, the spawning process must adhere to those limits. For example, if the boundaries are (-0.5, 0.5) for both x and y, the spawning process should respect these limits to ensure that entities remain within reach of one another.

In summary, the issue I'm explaining concerns how `x_semidim` and `y_semidim` are utilized from the kwargs.

**Comment by matteobettini** [comment]
Yes, makes sense, but if we do it this way the kwargs should be named `world_spawning_x` and `world_spawning_y` and default to 1
```
self.world_spawning_x = kwargs.pop("world_spawning_x", 1)
self.world_spawning_y = kwargs.pop("world_spawning_y", 1)
```
these would replace the kwargs we just introduced in the previous pr #133

the error is not needed anymore

**Comment by Giovannibriglia** [comment]
> Yes, makes sense, but if we do it this way the kwargs should be named `world_spawning_x` and `world_spawning_y` and default to 1
> 
> ```
> self.world_spawning_x = kwargs.pop("world_spawning_x", 1)
> self.world_spawning_y = kwargs.pop("world_spawning_y", 1)
> ```
> 
> these would replace the kwargs we just introduced in the previous pr #133
> 
> the error is not needed anymore

Understood, however my concerns regard the introduction of `world_spawning_x` and `world_spawning_y` in the kwargs. Will we apply this new logic across all suitable scenarios or only in the navigation scenario? If it's only for navigation, it might not be so elegant having these two only for one scenario, what do you think about?

Also, regarding setting 1 as default, I agree with you, it was my mistake.

**Comment by matteobettini** [comment]
I don't think we should consider families of related scenarios. Each scenario has its own kwargs and we don't guarantee any similarity across scenarios.

For example, discovery and navigation are different tasks, I don't think we should try to uniform them. The dense reward structure in navigation makes it suitable to unlimited bounds, while discovery needs bounding

**Comment by Giovannibriglia** [comment]
> I don't think we should consider families of related scenarios. Each scenario has its own kwargs and we don't guarantee any similarity across scenarios.
> 
> For example, discovery and navigation are different tasks, I don't think we should try to uniform them. The dense reward structure in navigation makes it suitable to unlimited bounds, while discovery needs bounding

Perfect!! Now it sounds good :)

File: vmas/scenarios/navigation.py [hunk 1]
```diff
@@ -24,8 +24,20 @@ def make_world(self, batch_dim: int, device: torch.device, **kwargs):
...
-        self.y_semidim = kwargs.pop("y_semidim", None)
+        self.world_spawning_x = kwargs.pop("world_spawning_x", 1)
+        self.world_spawning_y = kwargs.pop("world_spawning_y", 1)
+        self.enforce_bounds = kwargs.pop("enforce_bounds", False)
+
+        if self.enforce_bounds:
+            if self.world_spawning_x is None or self.world_spawning_y is None:
+                raise ValueError(
+                    "enforce_bounds is True, but one or both world_spawning variables are None."
+                )
```
Comment (matteobettini) [inline vmas/scenarios/navigation.py:35]:
```suggestion
```

this cannot happen anymore

File: vmas/scenarios/navigation.py [hunk 2]
```diff
@@ -296,6 +307,26 @@ def extra_render(self, env_index: int = 0) -> "List[Geom]":
...
+
+        # Set world spawning boundaries
+        self.world_spawning_x = (
+            self.x_semidim if self.enforce_bounds and self.x_semidim is not None else 1
+        )
+        self.world_spawning_y = (
+            self.y_semidim if self.enforce_bounds and self.y_semidim is not None else 1
+        )
+
```
Comment (matteobettini) [inline vmas/scenarios/navigation.py:329]:
```suggestion
```

**Comment by Giovannibriglia** [comment]
sorry, I forgot them :/

File: vmas/scenarios/navigation.py [hunk 3]
```diff
@@ -24,8 +24,16 @@ def make_world(self, batch_dim: int, device: torch.device, **kwargs):
...
+        self.world_spawning_x = kwargs.pop("world_spawning_x", 1)
+        self.world_spawning_y = kwargs.pop("world_spawning_y", 1)
+        self.enforce_bounds = kwargs.pop("enforce_bounds", False)
+
+        if self.enforce_bounds:
+            self.x_semidim = self.world_spawning_x
+            self.y_semidim = self.world_spawning_y
+        else:
+            self.x_semidim = None
+            self.y_semidim = None
```
Comment (matteobettini) [inline vmas/scenarios/navigation.py:36]: Could we move this after the kwargs reading?

Also maybe a small comment after each of the 3 new kwargs would help

Comment (Giovannibriglia) [reply vmas/scenarios/navigation.py]: Yes, it definitely helps. I tried to fix this in my latest commit :)

**Comment by matteobettini** [comment]
Thanks a lot, we should be good!
